proposals/dictionary-expressions.md

```diff
@@ -165,47 +165,47 @@ From [Open Question 3](#open-question-3), we will support *dictionary elements*
 2. Use applicable instance indexer if available; otherwise report an error during construction (or conversion?).
 3. Use C#12 initialization always.
 
+## Dictionary types
+
+The following are considered *dictionary types*:
+
+* A *type* with an appropriate *[create method](#create-methods)* where:
+  * The *iteration type* is `KeyValuePair<TKey, TValue>` determined from a `GetEnumerator` instance method or enumerable interface.
+  * The *type* has an instance *indexer* where:
+    * The indexer has a single parameter with an identity conversion from the parameter type to `TKey`.
+    * There is an identity conversion from the indexer type to `TValue`.
+    * The getter is `public`.
+* A *struct* or *class type* that implements `System.Collections.IEnumerable` where:
+  * The *iteration type* is `KeyValuePair<TKey, TValue>` determined from a `GetEnumerator` instance method or enumerable interface.
+  * The *type* has an *[applicable](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11642-applicable-function-member)* constructor that can be invoked with no arguments, and the constructor is accessible at the location of the collection expression.
+  * The *type* has an instance *indexer* where:
```

Comment:
> The *type* has an instance *indexer* where:

this looks repeated in this section.  It also seems like it would apply to the interface cases below.  Should we pull this out as a top level bullet item that all dictionary types have?

it seems like the only deviation is that we'd say we'd need a public-setter for the struct/class section.

Reply:
The conversion requirements for the indexer are distinct between the three cases:
- For a *type* with an appropriate *create method*, an indexer with a public getter is required, but not a setter.
- For a *struct* or *class type* that implements `IEnumerable`, an indexer with public getter and setter is required.
- For `IDictionary<TKey, TValue>` and `IReadOnlyDictionary<TKey, TValue>`, an indexer is not required for the conversion.

Reply:
> For a *type* with an appropriate *create method*, an indexer with a public getter is required, but not a setter.

Actually, the conversion requirement for an indexer could be dropped from this case too.

Reply:
sure, we could make all the cases distinct.  My point though is that perhaps it is cleaner/nicer to just say, uniformly, a dictionary type must have at least a `TValue this[TKey key] { get; }` indexer.

IMO, it's beneficial and means we can always defer to that member, uniformly, as a source of truth about the dictionary-ness of the type.

Reply:
Added a common indexer requirement.